### PR TITLE
hookstate/ctlcmd: snapctl unset command

### PR DESCRIPTION
**Pull request by stolowski**
Support `snapctl unset` command for removing config options from hooks.

### CODE REVIEW

File: overlord/hookstate/ctlcmd/unset_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	tr = config.NewTransaction(s.mockContext.State())
+	c.Check(tr.Get("test-snap", "foo", &value), ErrorMatches, `snap "test-snap" has no "foo" configuration option`)
+	c.Check(tr.Get("test-snap", "bar", &value), ErrorMatches, `snap "test-snap" has no "bar" configuration option`)
+	c.Check(tr.Get("test-snap", "baz", &value), IsNil)
+	c.Check(value, Equals, "c")
+}
+
+func (s *unsetSuite) TestUnsetRegularUserForbidden(c *C) {
+	state := state.New(nil)
+	state.Lock()
```
Comment (mvo5) [inline overlord/hookstate/ctlcmd/unset_test.go]: (nitpick^2) the \n here is a bit unusually (to me), I would put it more like:
```
state := ...

state.Lock()
do_stuff()
state.Unlock()

...
```
but not really an issue.

File: overlord/hookstate/ctlcmd/unset_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	s.mockContext, err = hookstate.NewContext(task, task.State(), setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+}
+
+func (s *unsetSuite) TestInvalidArguments(c *C) {
+	_, _, err := ctlcmd.Run(s.mockContext, []string{"unset"}, 0)
+	c.Check(err, ErrorMatches, "unset which option.*")
+}
+
+func (s *unsetSuite) TestUnsetMany(c *C) {
```
Comment (mvo5) [inline overlord/hookstate/ctlcmd/unset_test.go]: Should we have a test that checks for a single value unset too?

File: overlord/hookstate/ctlcmd/unset_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+	task := state.NewTask("test-task", "")
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "hook"}
+
+	state.Unlock()
+
+	mockHandler := hooktest.NewMockHandler()
+	mockContext, err := hookstate.NewContext(task, task.State(), setup, mockHandler, "")
+	c.Assert(err, IsNil)
+	_, _, err = ctlcmd.Run(mockContext, []string{"unset", "key"}, 1000)
```
Comment (mvo5) [inline overlord/hookstate/ctlcmd/unset_test.go]: Could we reuse s.mockContext here? It seems we don't do anything different there?

File: overlord/hookstate/ctlcmd/unset_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	task := state.NewTask("test-task", "")
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "hook"}
+
+	state.Unlock()
+
+	mockHandler := hooktest.NewMockHandler()
+	mockContext, err := hookstate.NewContext(task, task.State(), setup, mockHandler, "")
+	c.Assert(err, IsNil)
+	_, _, err = ctlcmd.Run(mockContext, []string{"unset", "-h"}, 1000)
```
Comment (mvo5) [inline overlord/hookstate/ctlcmd/unset_test.go]: Same question as above about reusing s.mockContext

File: overlord/hookstate/ctlcmd/unset_test.go (hunk 1 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset_test.go]: Good point.  I removed it alltogether though, it was a duplication of SetUpTest after all...

File: overlord/hookstate/ctlcmd/unset_test.go (hunk 2 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset_test.go]: Definately no harm in having it, added.

File: overlord/hookstate/ctlcmd/unset_test.go (hunk 3 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset_test.go]: Good point, removed. Thanks!

File: overlord/hookstate/ctlcmd/unset_test.go (hunk 4 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset_test.go]: Yep, removed as well.

File: overlord/hookstate/ctlcmd/unset.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	context := s.context()
+	if context == nil {
+		return fmt.Errorf("cannot unset without a context")
+	}
+
+	context.Lock()
+	tr := configstate.ContextTransaction(context)
+	context.Unlock()
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/unset.go]: I'm not entirely sure what lock/unlock protects against but should the rest of the transaction be protected by the lock? Did you mean to `defer context.Unlock()` here?

File: overlord/hookstate/ctlcmd/unset.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+var shortUnsetHelp = i18n.G("Remove configuration options")
+var longUnsetHelp = i18n.G(`
+The unset command removes the provided configuration options as requested.
+
+$ snapctl unset name address
+
+All configuration changes are persisted at once, and only after the
+snap's configuration hook returns successfully.
+
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset.go]: This is implied by the nature of hooks that are executed in the context of their snap and the help string is consistent with the rest of `snapctl` commands. Maybe rewording it to be more explicit is a good idea, but I'd reserve this for a followup where all commands (get/set/unset) would be updated at once.

File: overlord/hookstate/ctlcmd/unset.go (hunk 5 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/unset.go]: `configstate.ContextTransaction(context)` needs a context lock because it does `Get(..)` on the state to take a snapshot of snap's config. Lock is not needed afterwards for individual transaction `Set`s, it is acquired again only on `Commit` (which happens outside).
